filecoin-proofs/tests/paramfetch/support/session.rs

```diff
@@ -0,0 +1,121 @@
+extern crate rexpect;
+
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use failure::SyncFailure;
+use rexpect::session::PtyBashSession;
+use tempfile;
+use tempfile::TempDir;
+
+use crate::support::{cargo_bin, spawn_bash_with_retries};
+use storage_proofs::parameter_cache::PARAMETER_CACHE_ENV_VAR;
+
+pub struct ParamFetchSessionBuilder {
+    cache_dir: TempDir,
+    session_timeout_ms: u64,
+    whitelisted_sector_sizes: Option<Vec<String>>,
+    manifest: PathBuf,
+    prompt_enabled: bool,
+}
+
+impl ParamFetchSessionBuilder {
+    pub fn new(manifest: PathBuf) -> ParamFetchSessionBuilder {
+        let temp_dir = tempfile::tempdir().expect("could not create temp dir");
+
+        ParamFetchSessionBuilder {
+            cache_dir: temp_dir,
+            session_timeout_ms: 1000,
+            manifest,
+            prompt_enabled: true,
+            whitelisted_sector_sizes: None,
+        }
+    }
+
+    /// Configure the pty timeout (see documentation for `rexpect::spawn_bash`).
+    pub fn with_session_timeout_ms(mut self, timeout_ms: u64) -> ParamFetchSessionBuilder {
+        self.session_timeout_ms = timeout_ms;
+        self
+    }
+
+    /// Configure the pty timeout (see documentation for `rexpect::spawn_bash`).
+    pub fn whitelisted_sector_sizes(
+        mut self,
+        sector_sizes: Vec<String>,
+    ) -> ParamFetchSessionBuilder {
+        self.whitelisted_sector_sizes = Some(sector_sizes);
+        self
+    }
+
+    /// Create a file with the provided bytes in the cache directory.
+    pub fn with_file_and_bytes<P: AsRef<Path>, R: Read>(
+        self,
+        filename: P,
+        r: &mut R,
+    ) -> ParamFetchSessionBuilder {
+        let mut pbuf = self.cache_dir.path().clone().to_path_buf();
+        pbuf.push(filename.as_ref());
+
+        let mut file = File::create(&pbuf).expect("failed to create file in temp dir");
+
+        std::io::copy(r, &mut file).expect("failed to copy bytes to file");
+
+        self
+    }
+
+    /// Launch paramfetch in an environment configured by the builder.
+    pub fn build(self) -> ParamFetchSession {
```

Comment:
any reason for implementing this manually, rather than using sth like this: https://docs.rs/assert_cmd ?

Reply:
Indeed there is! These programs are interactive; they write a prompt to stdout and then block, waiting for user input. `assert_cmd` is a great tool for testing... commands. Unfortunately for me, `parampublish` and `paramfetch` behave more like a REPL than, say, `echo` or `ipfs`. This requires a pseudoterminal.